sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/healthcare/HttpHealthcareApiClient.java

```diff
@@ -0,0 +1,452 @@
+
+    Messages.List baseRequest =
+        client
+            .projects()
+            .locations()
+            .datasets()
+            .hl7V2Stores()
+            .messages()
+            .list(hl7v2Store)
+            .setPageToken(pageToken);
+
+    if (Strings.isNullOrEmpty(filter)) {
+      return baseRequest.execute();
+    } else {
+      return baseRequest.setFilter(filter).execute();
+    }
+  }
+
+  /**
+   * Gets message id page iterator.
+   *
+   * @param hl7v2Store the HL7v2 store
+   * @return the message id page iterator
+   * @throws IOException the io exception
+   */
+  @Override
+  public Stream<String> getHL7v2MessageIDStream(String hl7v2Store) throws IOException {
+    return getHL7v2MessageIDStream(hl7v2Store, null);
+  }
+
+  /**
+   * Get a {@link Stream} of message IDs from flattening the pages of a new {@link
+   * HL7v2MessageIDPages}.
+   *
+   * @param hl7v2Store the HL7v2 store
+   * @param filter the filter
+   * @return the message id Stream
+   * @throws IOException the io exception
+   */
+  @Override
+  public Stream<String> getHL7v2MessageIDStream(String hl7v2Store, @Nullable String filter)
+      throws IOException {
+    Iterator<List<String>> iterator = new HL7v2MessageIDPages(this, hl7v2Store, filter).iterator();
+    return flattenIteratorCollectionsToStream(iterator);
+  }
+
+  /**
+   * Gets HL7v2 message.
+   *
+   * @param msgName the msg name
+   * @return the message
+   * @throws IOException the io exception
+   * @throws ParseException the parse exception
+   */
+  @Override
+  public Message getHL7v2Message(String msgName) throws IOException {
+    Message msg =
+        client.projects().locations().datasets().hl7V2Stores().messages().get(msgName).execute();
+    if (msg == null) {
+      throw new IOException(String.format("Couldn't find message: %s.", msgName));
+    }
+    return msg;
+  }
+
+  @Override
+  public Empty deleteHL7v2Message(String msgName) throws IOException {
+    return client
+        .projects()
+        .locations()
+        .datasets()
+        .hl7V2Stores()
+        .messages()
+        .delete(msgName)
+        .execute();
+  }
+
+  /**
+   * Gets HL7v2 store.
+   *
+   * @param storeName the store name
+   * @return the HL7v2 store
+   * @throws IOException the io exception
+   */
+  @Override
+  public Hl7V2Store getHL7v2Store(String storeName) throws IOException {
+    return client.projects().locations().datasets().hl7V2Stores().get(storeName).execute();
+  }
+
+  @Override
+  public IngestMessageResponse ingestHL7v2Message(String hl7v2Store, Message msg)
+      throws IOException {
+    IngestMessageRequest ingestMessageRequest = new IngestMessageRequest();
+    ingestMessageRequest.setMessage(msg);
+    return client
+        .projects()
+        .locations()
+        .datasets()
+        .hl7V2Stores()
+        .messages()
+        .ingest(hl7v2Store, ingestMessageRequest)
+        .execute();
+  }
```

Comment:
Does the client handle quota-based retries on 429 responses?  Or is left to the caller to handle retries?

Reply:
yes [here](https://github.com/apache/beam/pull/11151/files/cac03ec30c77e6bd45954e659b875d0378370892#diff-3ab3eba677d52f20f1a688b1017adf84R274) we extend [`RetryHttpRequestInitializer`](https://beam.apache.org/releases/javadoc/2.13.0/org/apache/beam/sdk/extensions/gcp/util/RetryHttpRequestInitializer.html)